File: scripts/serve.py
from http.server import ThreadingHTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
from urllib.parse import urlparse, parse_qs, unquote
import os
import sys

# Flush stdout so the startup banner appears immediately.
sys.stdout.reconfigure(line_buffering=True)

ROOT = Path(__file__).resolve().parents[1]
os.chdir(ROOT)

# Optional preview generation via cv2. If cv2 is not installed, the
# ?preview=1 endpoint returns 404 and the frontend falls back to
# fetch-with-progress only.
try:
    import cv2
    HAS_CV2 = True
except ImportError:
    HAS_CV2 = False

PREVIEW_CACHE = {}
PREVIEW_CACHE_MAX = 200
IMAGE_EXTENSIONS = {".png", ".jpg", ".jpeg", ".webp", ".gif"}
MIME_TYPES = {
    ".css": "text/css; charset=utf-8",
    ".gif": "image/gif",
    ".html": "text/html; charset=utf-8",
    ".ico": "image/x-icon",
    ".jpeg": "image/jpeg",
    ".jpg": "image/jpeg",
    ".js": "text/javascript; charset=utf-8",
    ".json": "application/json; charset=utf-8",
    ".mjs": "text/javascript; charset=utf-8",
    ".png": "image/png",
    ".svg": "image/svg+xml",
    ".webp": "image/webp",
}


def generate_preview(file_path):
    """Generate a tiny low-quality JPEG preview (~48px wide)."""
    if not HAS_CV2:
        return None

    cached = PREVIEW_CACHE.get(file_path)
    if cached is not None:
        return cached

    try:
        img = cv2.imread(str(file_path), cv2.IMREAD_COLOR)
        if img is None:
            return None
        height, width = img.shape[:2]
        scale = 48.0 / width if width > 48 else 1.0
        if scale < 1.0:
            new_width = 48
            new_height = max(1, int(height * scale))
            img = cv2.resize(img, (new_width, new_height), interpolation=cv2.INTER_AREA)

        ok, buffer = cv2.imencode(".jpg", img, [cv2.IMWRITE_JPEG_QUALITY, 32])
        if not ok:
            return None

        if len(PREVIEW_CACHE) >= PREVIEW_CACHE_MAX:
            PREVIEW_CACHE.pop(next(iter(PREVIEW_CACHE)))
        PREVIEW_CACHE[file_path] = buffer.tobytes()
        return PREVIEW_CACHE[file_path]
    except Exception:
        return None


class CAGERequestHandler(SimpleHTTPRequestHandler):
    def end_headers(self):
        # Prevent aggressive browser caching of JS/CSS during development.
        path = getattr(self, "_resolved_suffix", "")
        if path in (".js", ".css", ".mjs"):
            self.send_header("Cache-Control", "no-cache, must-revalidate")
        super().end_headers()

    def do_GET(self):
        parsed = urlparse(self.path)
        query = parse_qs(parsed.query)
        relative_path = unquote(parsed.path.lstrip("/"))
        file_path = (ROOT / relative_path).resolve()

        # Security: prevent path traversal.
        try:
            file_path.relative_to(ROOT)
        except ValueError:
            self.send_error(403, "Forbidden")
            return

        # Serve index.html for directory requests.
        if file_path.is_dir():
            file_path = file_path / "index.html"

        if not file_path.is_file():
            self.send_error(404, "Not found")
            return

        ext = file_path.suffix.lower()
        is_image = ext in IMAGE_EXTENSIONS

        # Preview endpoint: ?preview=1 on an image returns a tiny JPEG.
        if "preview" in query and is_image:
            preview = generate_preview(file_path)
            if preview is None:
                self.send_error(404, "Preview not available")
                return
            self.send_response(200)
            self.send_header("Content-Type", "image/jpeg")
            self.send_header("Cache-Control", "public, max-age=86400")
            self.send_header("Content-Length", str(len(preview)))
            self.end_headers()
            self.wfile.write(preview)
            return

        if is_image:
            stat = file_path.stat()
            self.send_response(200)
            self.send_header("Content-Type", MIME_TYPES.get(ext, "application/octet-stream"))
            self.send_header("Cache-Control", "public, max-age=3600")
            self.send_header("Content-Length", str(stat.st_size))
            self.end_headers()
            with open(file_path, "rb") as f:
                while True:
                    chunk = f.read(65536)
                    if not chunk:
                        break
                    self.wfile.write(chunk)
            return

        # Fallback for non-image files (HTML, CSS, JS, etc.).
        self._resolved_suffix = ext
        super().do_GET()


if __name__ == "__main__":
    print(f"Serving CAGE at http://127.0.0.1:8000")
    if HAS_CV2:
        print("Image preview generation enabled (cv2).")
    else:
        print("Image preview disabled. Install opencv-python for progressive loading: pip install opencv-python")
    ThreadingHTTPServer(("127.0.0.1", 8000), CAGERequestHandler).serve_forever()

File: scripts/serve.mjs
import { createReadStream, statSync } from "node:fs";
import { createServer } from "node:http";
import { extname, resolve, sep } from "node:path";
import { fileURLToPath } from "node:url";

const root = resolve(fileURLToPath(new URL("..", import.meta.url)));
const port = Number(process.env.PORT || 8000);

// Optional image processing for low-quality preview generation.
// The server works fine without sharp — previews are simply disabled
// and the frontend falls back to fetch-with-progress only.
let sharp = null;
try {
    sharp = (await import("sharp")).default;
} catch (_) {
    // sharp is not installed — preview generation disabled.
}

const mimeTypes = {
    ".css": "text/css; charset=utf-8",
    ".gif": "image/gif",
    ".html": "text/html; charset=utf-8",
    ".ico": "image/x-icon",
    ".jpeg": "image/jpeg",
    ".jpg": "image/jpeg",
    ".js": "text/javascript; charset=utf-8",
    ".json": "application/json; charset=utf-8",
    ".mjs": "text/javascript; charset=utf-8",
    ".png": "image/png",
    ".svg": "image/svg+xml",
    ".webp": "image/webp",
};

const IMAGE_EXTENSIONS = new Set([".png", ".jpg", ".jpeg", ".webp", ".gif"]);

// In-memory cache for generated previews so repeated requests are instant.
const previewCache = new Map();
const PREVIEW_CACHE_MAX = 200;

async function generatePreview(filePath) {
    if (!sharp) return null;

    const cached = previewCache.get(filePath);
    if (cached) return cached;

    try {
        // Resize to ~48px wide (maintaining aspect ratio) and encode as a
        // low-quality progressive JPEG. This produces a ~1-3 KB preview that
        // the browser can download almost instantly and display as a blurry
        // placeholder while the full-resolution image continues loading.
        const buffer = await sharp(filePath)
            .resize(48, null, { withoutEnlargement: true })
            .jpeg({ quality: 32, progressive: true })
            .toBuffer();

        if (previewCache.size >= PREVIEW_CACHE_MAX) {
            const firstKey = previewCache.keys().next().value;
            previewCache.delete(firstKey);
        }
        previewCache.set(filePath, buffer);
        return buffer;
    } catch (_) {
        return null;
    }
}

createServer(async (request, response) => {
    try {
        const requestUrl = new URL(request.url || "/", "http://localhost");
        const pathname = decodeURIComponent(requestUrl.pathname);
        const relativePath = pathname === "/" ? "index.html" : pathname.replace(/^\/+/, "");
        let filePath = resolve(root, relativePath);

        if (filePath !== root && !filePath.startsWith(root + sep)) {
            response.writeHead(403).end("Forbidden");
            return;
        }

        const stats = statSync(filePath);
        if (stats.isDirectory()) {
            filePath = resolve(filePath, "index.html");
        }

        const ext = extname(filePath).toLowerCase();
        const isImage = IMAGE_EXTENSIONS.has(ext);

        // Low-quality preview endpoint: ?preview=1 on an image path returns a
        // tiny progressive JPEG. Returns 404 when sharp is unavailable so the
        // frontend can detect this and skip the preview phase.
        if (requestUrl.searchParams.has("preview") && isImage) {
            if (!sharp) {
                response.writeHead(404, { "Content-Type": "text/plain" });
                response.end("Preview not available");
                return;
            }
            const previewBuffer = await generatePreview(filePath);
            if (!previewBuffer) {
                response.writeHead(500, { "Content-Type": "text/plain" });
                response.end("Preview generation failed");
                return;
            }
            response.writeHead(200, {
                "Content-Type": "image/jpeg",
                "Cache-Control": "public, max-age=86400",
                "Content-Length": previewBuffer.length,
            });
            response.end(previewBuffer);
            return;
        }

        const headers = {
            "Content-Type": mimeTypes[ext] || "application/octet-stream",
            "Cache-Control": isImage ? "public, max-age=3600" : "no-cache",
        };
        response.writeHead(200, headers);
        createReadStream(filePath).pipe(response);
    } catch (error) {
        response.writeHead(404, { "Content-Type": "text/plain; charset=utf-8" });
        response.end("Not found");
    }
}).listen(port, "127.0.0.1", () => {
    console.log(`Serving CAGE at http://127.0.0.1:${port}`);
    if (sharp) {
        console.log("Image preview generation enabled (sharp).");
    } else {
        console.log("Image preview disabled. Install sharp for progressive loading: npm install sharp");
    }
});
